Adding AWS Transcribe

edsu opened the pull request:
This commit adds a test for AWS Transcribe, and a (not yet implemented) test for Google Speech to Text. The existing whisper module (report.py) was moved to transcribe/whisper.py.

After some initial review an addition set of changes was added to this PR to get the new aws module reporting results similar to whisper. This required moving some code from whisper into a shared module `transcribe.utils`. I also added some new tests.

The `run` program to be able to run the full report, or individual ones. Results are also now written to a timestamped directory which can be changed with the `--output-dir` command line option.

edsu committed:
Adding aws and google

This commit adds a test for AWS Transcribe, and an (not yet
implemented) test for Google Speech to Text. The existing whisper module
(report.py) was moved to transcribe/whisper.py.

edsu commented on .github/workflows/test.yml:
```diff
@@ -20,6 +20,7 @@ jobs:
 
     - name: Install dependencies
       run: |
+        sudo apt update -y
```
I noticed that ffmpeg wasn't installing this afternoon without updating apt first.

edsu commented on .ruff.toml:
```diff
@@ -1 +1,2 @@
-ignore = ["ANN"]
+[lint]
+select = ["E4", "E7", "E9", "F", "I"]
```
These are rules for linting with [ruff](https://docs.astral.sh/ruff/).

edsu commented on env-example:
```diff
@@ -0,0 +1,9 @@
+#
+# AWS_ROLE_ARN=
+# AWS_REGION=
+# AWS_PROFILE=
```
The `.env` file is only needed if you are interacting with AWS or Google.

edsu commented on run.py:
```diff
@@ -0,0 +1,18 @@
+if args.preprocessing:
+    whisper.run_preprocessing()
+else:
+    whisper.run()
```
Since the `report.py` was moved to `transcribe/whisper.py` I moved the code to run things here. I also added argparse to parse whether to run preprocessing.

edsu commented on test/test_aws.py:
```diff
@@ -0,0 +1,20 @@
+    assert result == {
+        "language": "en-US",
+        "transcript": "This is a test for whisper reading in English.",
+    }
```
A new test for testing the basic AWS functionality.

edsu commented on test/test_google.py:
```diff
@@ -0,0 +1,19 @@
+    assert result == {
+        "language": "en-US",
+        "transcript": "This is a test for whisper reading in English.",
+    }
```
This test is really just testing a stubbed out interface.

edsu commented on test/test_whisper.py:
```diff
@@ -0,0 +1,33 @@
+    assert (
+        whisper.get_language(path.join(TEST_DATA, "fr-with-silence.wav"), MODEL_SIZE)
+        == "fr"
+    )
```
This was moved here from `test_report.py` it is largely the same except for some changes around where to load the data from.

edsu commented on test_report.py:
moved to `test/test_whisper.py`

edsu commented on transcribe/aws.py:
```diff
@@ -0,0 +1,92 @@
+    if job["TranscriptionJob"]["TranscriptionJobStatus"] == "COMPLETED":
+        return job
+    else:
+        return wait_for_job(scribe, job_name, wait_seconds**2)
```
This is new code for interacting with AWS Transcribe.

dnoneill replied:
I think it would be good if we could get all the same data that we are getting from whisper. So length of time it is taking to run and wer would also be good for these scripts to do.

edsu replied:
Yes I'm actually planning on pulling the timing and jiwere out of the Whisper module and then reusing it for each transcription. That will be another PR though?

edsu commented on transcribe/google.py:
```diff
@@ -0,0 +1,5 @@
+    return {
+        "language": "en_US",
+        "transcript": "This is a test for whisper reading in English.",
+    }
```
This is stubbed out code for interacting with Google's Speech to Text.

edsu commented on transcribe/whisper.py:
```diff
@@ -304,10 +303,3 @@ def load_model(model_name):
-    if len(sys.argv) > 1 and sys.argv[1] == "preprocessing":
-        run_preprocessing()
-    else:
-        run()
```
This was moved to `run.py`

edsu committed:
Created transcribe.utils

In order to share code for discovering media files, their transcripts,
and generating the report these functions in the whisper module were
moved into transcribe.utils.

The run script was also updated to run all transcription types by
default, or to allow them to be run individually with --only.

edsu commented on transcribe/whisper.py:
```diff
@@ -0,0 +1,243 @@
+@cache
+def load_model(model_name):
+    # cache the response since it takes some time to load
+    device = "cuda" if torch.cuda.is_available() else "cpu"
```
This is very important for getting Whisper to use the GPU on `li-rd-7217.stanford.edu`. Without this it simply uses the CPU.
